RFC: operation expressions

benjie opened the pull request:
Relates to https://github.com/graphql/graphql-wg/issues/592

This RFC is to act as a discussion place for operation expressions; an extension to the Schema Coordinates syntax that allows for a host of new use cases.

I've outlined some of the use cases that have been discussed so far, and have iterated on the syntax a little, but everything in this is very much in flux. The purpose of this PR is to be a place to host the discussion (separate from Schema Coordinates) so that others may bring their ideas to the table. I don't think it's suitable for merging at this time.

[See the document rendered nicely here](https://github.com/benjie/graphql-spec/blob/operation-expressions/rfcs/OperationExpressions.md).

Grammar support:

- [x] Anonymous operations
- [x] Named operations
- [x] Named fragments
- [x] Inline fragments
- [x] Named fragments within a path
- [x] Arguments
- [x] Argument nested input object references
- [x] Lists
- [ ] Directives

benjie commented:
cc @magicmark

sungam3r commented on rfcs/OperationExpressions.md:
```diff
@@ -0,0 +1,219 @@
+- `query>me>name` expands to `query { me { name } }`
+
+You may name operations by prefixing with an operation name followed by a
+colon; for example `MyName:>me>name` expands to `query MyName { me { name } }`.
```
What about named mutations/subscriptions ? `mutation.DoIt:>me>name` -> `mutation DoIt { me { name } }` Right? What character separates the operation name from its type?

benjie replied:
Same as with queries: prefix with name and colon: `MyMutation:mutation>createUser>username`

sungam3r replied:
An example is needed.

sungam3r commented on rfcs/OperationExpressions.md:
```diff
@@ -0,0 +1,219 @@
+
+#### Fragments
+
+Fragments start with a type name followed by a period: `User.friends>name`
```
How do you distinguish this from named operations? See https://github.com/graphql/graphql-spec/pull/823/files#r577833203

benjie replied:
Fragments start with a typename period fieldname, operations start with operation type (optionally) greater than, fieldname

fragment: `User.field` Or `Name:User.field`

operation: `query>field` `>field` `Name:mutation>field`

sungam3r commented on rfcs/OperationExpressions.md:
```diff
@@ -0,0 +1,219 @@
+
+We also allow you to reference input objects used in arguments, for example:
+
+`>searchBusinesses(where.size.greaterThan:)>city`
```
What if I have multiple arguments?

benjie replied:
Good point, plan is to use comma but I forgot to write that. (Figured using space would open a can of worms; disallowing all spaces seemed the smarter choice)

sungam3r replied:
I'm also interested in the question of how to set multiple fields in one input object (if applicable).

benjie replied:
I think something like: `>searchBusinesses(where.size.greaterThan:,where.size.lessThan:,where.city.equalTo:)>name`; what's your use case?

sungam3r replied:
Just interested.

benjie commented:
Thanks for the feedback @sungam3r; but I’m more interested in use cases and general reasons not to have a standard syntax like this rather than syntax specifics right now. We might replace any of the symbols or even change the grammar before this sees the light of day; the aim currently is to see if it’s worth investing time here.

eapache commented:
Great start! I have no concerns with the use cases so far.

michaelstaib commented:
@benjie this looks awesome :) really love the idea. All in for it.

benjie commented:
@sungam3r All feedback addressed; I've also gone as far as to write out the full grammar in spec-md format.

benjie commented:
Added to [the next WG agenda](https://github.com/graphql/graphql-wg/blob/main/agendas/2021-03-04.md#agenda); main aim right now is to collect more use cases.

The main motivation for doing this now is to make sure that the Schema Coordinates RFC syntax is extensible. So far, I feel that it is.

sungam3r commented:
@benjie From all your examples it follows that we are always talking about a single field. It was not clear to me at first, especially when I read about generated queries. I began to wonder how to make it possible to specify more than one field in the generated request. As I understand it, in no way, because this is not the goal. The goal is to give the coordinate of one field, and the generated request is an additional feature. Correct?

benjie commented:
@sungam3r I'm still gathering use cases, so if you have a use case for multiple fields I'd love to hear it and we could potentially expand the syntax to incorporate that. However in general for referencing multiple fields it feels like we already have a query syntax that handles that: GraphQL itself.

Currently all the examples in this PR relate to referencing a single field; but what I'd really like to see is all the different ways that we might build upon the syntax from the Schema Coordinates RFC. If there are use cases that require extensions to the Schema Coordinates syntax but don't relate to fields, I'd absolutely love to hear them! Don't let your imagination be limited by the few examples I've added, nor by the syntax I've added to address these examples.

magicmark commented on rfcs/OperationExpressions.md:
```diff
@@ -0,0 +1,391 @@
+
+## Use cases
+
+#### Referencing a position within a GraphQL Operation Document
```
pedantic question: should this be "Referencing a position within a GraphQL Query"?

asking cos what if the document contains two operations? does the operation expression encode which operation in the document it refers to? (can it / should it?)

benjie replied:
Yes it can, via the name:

```
OperationName:mutation>createFoo>foo>id
```

magicmark commented on rfcs/OperationExpressions.md:
```diff
@@ -0,0 +1,391 @@
+- `>businesses:searchBusinesses(name:)>owner:personByOwnerId>email` - this
+  contains even more context (the argument names that were used)
+
+These are all valid operation expressions, but they each convey different levels
+of context.
```
I really like this - i could imagine being able to specify different levels of verbosity to tooling that generates these

magicmark commented on rfcs/OperationExpressions.md:
```diff
@@ -0,0 +1,391 @@
+> Emmet is a plugin for many popular text editors which greatly improves HTML &
+> CSS workflow:
+
+Emmet is a popular syntax for quickly generating HTML/CSS. It's easy to imagine
+how a operation expression syntax could be combined with a GraphQL schema
+definition to quickly generate GraphQL queries, mutations and subscriptions with
+a concise syntax. For example the expression:
```
How lossy is this? If I converted a query to an operation expression and then back again to a query, is it always perfectly preserved? (Wondering how things like @if are preserved, or directives in general)

benjie replied:
Directives are out of scope currently (though I can imagine them being added, repeatable may be an awkward syntax) and it currently only maps one field/entity/path so cannot contain multiple fields like a query would.

magicmark commented on rfcs/OperationExpressions.md:
```diff
@@ -0,0 +1,391 @@
+We also allow `[number]` syntax to refer to a numbered entry in a list, or `[]`
+to refer to the next entry; e.g.:
+
+`>findUsers(byIds[]:,byIds[],byIds[],byIds[5])>name`
```
i'm a little confused here - are we implying that `byIds5` is an array?

(it's writen as a single `ID` type in the query below)

benjie replied:
The auto-variables may be adding some confusion here. The `byIds:` _argument_ is a list, so we can specify members of it; since they have no values if we want to have it converted to a document we need to pass something so we automatically generate variables for it; in this case `$byIds5` (an ID, not a list) is one of the members of the list that’s fed to `byIds:`

magicmark reviewed (commented):
@benjie this looks amazing! Thanks for this huge effort (as always :P)

Wondering if we should add a non-goals or open questions section to clarify a couple of thoughts (and I imagine others) had:

- Should this syntax support selecting multiple leaf nodes with commas? Wildcard selectors?
- Should this take directives into account? (I'm having a hard time imagining what the use case here would be / what that looks like)

Other than that, the direction/syntax here looks good to me :)

magicmark commented on rfcs/OperationExpressions.md:
```diff
@@ -0,0 +1,391 @@
+
+Further we allow for multiple arguments to be specified, joined with commas:
+
+`>searchBusinesses(where.size.greaterThan:,where.size.lessThan:,where.city.equalTo:)>name`
```
I notice we're using a dot separator for input types (`where.size.greaterThan`)

any reason we're opting for . instead of > here?

e.g. this:

`>searchBusinesses(where>size>greaterThan:,where>size>lessThan:,where>city>equalTo:)>name`

(even if the answer is "it looks nicer", then fair enough, just wondering)

benjie replied:
Good point. Earlier syntax didn’t have >, I had to add it to avoid ambiguity and didn’t think to update the arguments. May be worth adopting as you say to avoid issues with input unions or similar in future.

benjie replied:
Done :+1:

magicmark commented:
I see enums are described as supported, but not called out as an example or in the list of grammar support. Should there be an example?

benjie commented:
@magicmark I can't see the term `enum` in the document other than referencing your Schema Coordinates RFC. I can't think of a use case that needs enum support right now; do you have one in mind?

leebyron commented:
Merging in advance of moving the RFCs to a new repo
